pyiron/atomistics/structure/atoms.py

```diff
@@ -1744,120 +1688,52 @@ def _get_neighbors(
-            neighbors = tree.query(positions, k=num_neighbors)
+            prefactor = [1, 1/np.pi, 4/(3*np.pi)]
+            prefactor = prefactor[sum(self.pbc)-1]
+            width = np.prod((np.linalg.norm(self.cell, axis=-1)-np.ones(3))*self.pbc+np.ones(3))
+            width *= prefactor*np.max([num_neighbors, 8])/len(self)
+            width = boundary_width_factor*width**(1/np.sum(self.pbc))
+        neighbor_obj._boundary_layer_width = width
+        if (width<0.5*np.min(self.cell.diagonal())
+            and np.isclose(np.linalg.norm(self.cell-np.eye(3)*self.cell.diagonal()), 0)
+            and np.all(self.pbc)
+            and cutoff_radius==np.inf):
+            boxsize = self.cell.diagonal()
+            extended_positions, indices = self.get_extended_positions(0)
+            extended_positions /= self.cell.diagonal()
+            extended_positions -= np.floor(extended_positions)
+            extended_positions *= self.cell.diagonal()
         else:
-            neighbors = tree.query(
-                positions, k=num_neighbors, distance_upper_bound=cutoff_radius
-            )
-
-        # print ("neighbors: ", neighbors)
-
+            extended_positions, indices = self.get_extended_positions(width)
+        if len(extended_positions) < num_neighbors and cutoff_radius==np.inf:
+            raise ValueError('num_neighbors too large - make boundary_width_factor larger and/or make num_neighbors smaller')
+        tree = cKDTree(extended_positions, boxsize=boxsize)
+        if id_list is None:
+            id_list = np.arange(len(self.positions))
+        positions = self.positions[np.array(id_list)]
+        neighbors = tree.query(
+            positions, k=num_neighbors, distance_upper_bound=cutoff_radius
+        )
         neighbor_distance = []  # neighbors[0]
         neighbor_distance_vec = []
         neighbor_indices = []
-
-        # tolerance = 2 # tolerance for round floating point
-
-        def f_ind_ext(x):
-            return x < len(extended_cell)
-
-        neighbor_index = map(lambda x: filter(f_ind_ext, x), neighbors[1])
-        num_neighbors = []
-        for i, index in enumerate(neighbor_index):
-            # print "i, index: ", i, index
-            index = list(index)  # Filter conversion for python 3 compatibility
-            nbrs_distances = neighbors[0][i][i_start : len(index)]
-            # if radius:  # reduce neighborlist based on radius
-            #     new_index_lst, new_dist_lst = [], []
-            #     for index_red, dis_red in zip(index, nbrs_distances):
-            #         if dis_red < radius:
-            #             new_index_lst.append(index_red)
-            #             new_dist_lst.append(dis_red)
-            #     index, nbrs_distances= new_index_lst, new_dist_lst
-            neighbor_distance.append(nbrs_distances)
-            neighbor_indices.append(map_to_cell[index][i_start:])
-
+        for atom_id, index, distances in zip(id_list, neighbors[1], neighbors[0]):
```

Comment:
```python
n = min(len(neighbors[1], neighbors[2])
neighbor_distance = np.zeros(n)
neighbor_distance_vec = np.zeros(n)
neighbor_indices = np.zeros(n)
for atom_id, (index, distances) in enumerate(zip(neighbors[1], neighbors[2]):
     neighbor_distance[atom_id] = distances[1:][index[1:]<len(indices)]
     ...
if id_list is not None:
    neighbor_indices =np.asarray(id_list)[neighbor_indices]
```
or something similar would avoid repeatably appending to python lists, which is very costly usually.

Reply:
list append is not slow in contrast to np append or common C append.

Reply:
But ok I can change it